src/nb_type.cpp

```diff
@@ -418,17 +434,43 @@ PyObject *nb_type_new(const type_data *t) noexcept {
             has_traverse = true;
     }
 
-    if (has_dynamic_attr) {
-        // realign to sizeof(void*), add one pointer
+    Py_ssize_t dictoffset = 0;
+    Py_ssize_t weaklistoffset = 0;
+    int num_members = 0;
+    if (has_dynamic_attr || is_weak_referenceable) {
+        // realign to sizeof(void*)
         basicsize = (basicsize + ptr_size - 1) / ptr_size * ptr_size;
+    }
+    if (has_dynamic_attr) {
+        dictoffset = (Py_ssize_t) basicsize;
         basicsize += ptr_size;
 
-        members[0] = PyMemberDef{ "__dictoffset__", T_PYSSIZET,
-                                  (Py_ssize_t) (basicsize - ptr_size), READONLY,
-                                  nullptr };
-        *s++ = { Py_tp_members, (void *) members };
+        members[num_members] = PyMemberDef{ "__dictoffset__", T_PYSSIZET,
+                                            dictoffset, READONLY,
+                                            nullptr };
+        ++num_members;
+
+        // Install GC traverse and clear routines if not inherited/overridden
+        if (!has_traverse) {
+            *s++ = { Py_tp_traverse, (void *) inst_traverse };
+            *s++ = { Py_tp_clear, (void *) inst_clear };
+            has_traverse = true;
+        }
+        spec.basicsize = (int) basicsize;
+    }
+    if (is_weak_referenceable) {
+        weaklistoffset = (Py_ssize_t) basicsize;
+        basicsize += ptr_size;
+
+        members[num_members] = PyMemberDef{ "__weaklistoffset__", T_PYSSIZET,
+                                            weaklistoffset, READONLY,
+                                            nullptr };
+        ++num_members;
 
         // Install GC traverse and clear routines if not inherited/overridden
+        // Technically having weak references does not imply we need GC. However
```

Comment:
I'm not entirely happy with this, but I'm trying to avoid the direct call to `PyObject_Malloc()` in inst_new_impl which calculates its own size and wouldn't get the right size. This seemed like it was the simplest fix.

Reply:
It's fine with me. The cost of enabling GC for a type is not that high. The requirement of a weakly-reference type is sufficiently narrow that this seems like the easiest compromise.